src/units.cpp

```diff
@@ -457,43 +459,99 @@ void updateUnitsMap(const UnitsPtr &units, UnitsMap &unitsMap, double exp = 1.0)
                 if (model != nullptr) {
                     auto refUnits = model->units(ref);
                     if ((refUnits == nullptr) || refUnits->isImport()) {
-                        unitsMap.clear();
-                        break;
+                        return false;
+                    }
+                    if (!updateUnitsMap(refUnits, unitsMap, uExp * exp)) {
+                        return false;
                     }
-                    updateUnitsMap(refUnits, unitsMap, uExp * exp);
                 }
             }
         }
     }
+    return true;
 }
 
-UnitsMap createUnitsMap(const UnitsPtr &units)
+UnitsMap createUnitsMap(const UnitsPtr &units, bool &isValid)
 {
     UnitsMap unitsMap;
-    updateUnitsMap(units, unitsMap);
+    isValid = true;
+    if (!updateUnitsMap(units, unitsMap)) {
+        isValid = false;
+        return unitsMap;
+    }
 
     // Checking for exponents of zero in the map, which can be removed.
-    bool requireDimensionless = false;
     auto it = unitsMap.begin();
     while (it != unitsMap.end()) {
         if (it->second == 0.0) {
             it = unitsMap.erase(it);
-            requireDimensionless = true;
-        } else if (it->first == "dimensionless") {
-            it->second = 0.0;
-            ++it;
+        } else if ((it->first == "dimensionless") || (it->first == "radian") || (it->first == "steradian")) {
```

Comment:
Does this mean we have to know the names of all `dimensionless` units?

Reply:
Unless you're happy to say that `radians != dimensionless`, then yes, we need to specify them here.  I wasn't sure about this (see #552)...

Reply:
Scratch that, I think you're right, thanks :)